### DISCUSSION TITLE
Connecting Cassandra to JNDI in WildFly: A Guide for Integration with cassandra-jdbc-wrapper-4.11.0.jar

### DISCUSSION THREAD

**Discussion by Arjunguptha98** [ing-bank/cassandra-jdbc-wrapper#47] [Q&A]
Hi team,
I've developed this project and obtained the cassandra-jdbc-wrapper-4.11.0.jar. With this JAR, I've successfully executed a Java JDBC program. Currently, I'm attempting to establish a connection between Cassandra and JNDI (using WildFly). Could you provide guidance on achieving this use case if it's possible?

**Comment by maximevw** [comment]
Hello @Arjunguptha98,

I´m not sure to understand your question.
If you want to configure a JDBC datasource identified by a JNDI name in WildFly, there is nothing specific to do for this driver and you can configure the datasource as any JDBC datasource in Wildfly as explained in the [reference documentation](https://docs.wildfly.org/30/Admin_Guide.html#DataSource).

And you can refer to [this page](https://github.com/ing-bank/cassandra-jdbc-wrapper/wiki/JDBC-driver-and-connection-string) for driver and connection URL settings.

**Comment by Arjunguptha98** [comment]
Hi @maximevw 
Thank you for your reply. I've obtained the JDBC URL for Cassandra. However, I'm uncertain about how to configure the Cassandra driver in standalone.xml and module.xml.

**Comment by maximevw** [comment]
Hello @Arjunguptha98 

First of all, I will convert this to a discussion as your question is not an issue related to the JDBC driver. :)

I have no expertise with Wildfly, but based on the following sources ([here](https://medium.com/@tiagomac/configuring-a-datasource-in-wildfly-and-connecting-to-it-through-spring-boot-8c2ef880670b) and [there](https://tomylab.wordpress.com/2016/07/24/how-to-add-a-datasource-to-wildfly/)), I think you should do something like this:

**(UPDATED on base of the next comments)**

1. Download the [driver JAR](https://github.com/ing-bank/cassandra-jdbc-wrapper/releases/download/v4.11.0/cassandra-jdbc-wrapper-4.11.0-bundle.jar).
2. Put the JAR in the folder `{WildFlyDirectory}/modules/system/layers/base/com/ing/data/main` and create a module.xml file in the same folder:
```xml
<module xmlns="urn:jboss:module:1.9" name="com.ing.data">
  <resources>
    <resource-root path="cassandra-jdbc-wrapper-4.11.0-bundle.jar"/>
  </resources>
  <dependencies>
    <module name="javax.api"/>
    <module name="javax.transaction.api"/>
    <module name="org.apache.logging.log4j.api"/>
  </dependencies>
</module>
```
3. In the standalone.xml file, locate the `<subsystem>/<datasources>` part and include the following datasource :
```xml
<datasource jndi-name="java:jboss/datasources/cassandraDS" pool-name="cassandraDS" enabled="true">
  <connection-url>{JDBC URL}</connection-url>
  <driver>cassandra</driver>
  <security user-name="{username}" password="{password}" />
  </security>
</datasource>
```
4. Then, also in standalone.xml, locate the tag `<drivers>` and put this inside:
```xml
<driver name="cassandra" module="com.ing.data">
  <driver-class>com.ing.data.cassandra.jdbc.CassandraDriver</driver-class>
</driver>
```
5. Reboot the Wildfly server.

There is no guarantee it works exactly like this, I just deduced that from various documentation found on the Internet and I never tested this.

I really encourage you to consult the [WildFly Admin guide](https://docs.wildfly.org/30/Admin_Guide.html) for further details on JDBC datasource configuration.

**Comment by Arjunguptha98** [comment]
Hi @maximevw. Thanks for the response.
I attempted to configure Cassandra with Wildfly following the provided steps. However, I encountered a dependency issue, as outlined below.
![image](https://github.com/ing-bank/cassandra-jdbc-wrapper/assets/53363410/e7d68ab3-4c65-43c0-8f6e-43b0b12a784a)

**Comment by maximevw** [comment]
@Arjunguptha98, 

It seems it doesn't find the driver. At step 2, I think there is a mistake and the directory should be `{WildFlyDirectory}/modules/system/layers/base/com/ing/data/main`.

**Comment by Arjunguptha98** [comment]
Hi @maximevw .
Still getting same dependency issue

**Reply by maximevw** [reply]
@Arjunguptha98, I installed Wildfly 30.0.1.Final locally, without deploying any application. And I just configured the datasource as specified above. I just modified a little the configuration as following:
```xml
<datasource jndi-name="java:jboss/datasources/cassandraDS" pool-name="cassandraDS" enabled="true">
  <connection-url>{JDBC URL}</connection-url>
  <driver>cassandra</driver>
  <security user-name="{username}" password="{password}"/>
</datasource>
```

And it seems it works:
```
18:21:28,493 INFO  [org.jboss.as.connector.subsystems.datasources] (ServerService Thread Pool -- 44) WFLYJCA0005: Deploying non-JDBC-compliant driver class com.ing.data.cassandra.jdbc.CassandraDriver (version 4.11)
18:21:28,497 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-7) WFLYJCA0018: Started Driver service with driver-name = cassandra
...
18:21:29,009 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-8) WFLYJCA0001: Bound data source [java:jboss/datasources/ExampleDS]
18:21:29,016 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-5) WFLYJCA0001: Bound data source [java:jboss/datasources/cassandraDS]
...
18:21:29,282 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0212: Resuming server
18:21:29,285 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0060: Http management interface listening on http://127.0.0.1:9990/management
18:21:29,286 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0051: Admin console listening on http://127.0.0.1:9990
18:21:29,288 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 30.0.1.Final (WildFly Core 22.0.2.Final) started in 3466ms - Started 288 of 530 services (318 services are lazy, passive or on-demand) - Server configuration file in use: standalone.xml

```
And in the admin console, if I test the connection, it's ok.
![image](https://github.com/ing-bank/cassandra-jdbc-wrapper/assets/48218208/6ea2f73b-e2d7-4ca1-acd9-eefa7cbf5574)

Ensure the driver "cassandra" is present in the section `<drivers>` (just under the section `<datasource>`, inside the section `<datasources>`).

**Comment by Arjunguptha98** [comment]
Thank you for your insightful feedback. The details you shared were instrumental in resolving the issues I encountered with the setup.

**Comment by Arjunguptha98** [comment]
Hi @maximevw 
While trying test the connection of cassandraDS(Test Connection) from the UI. I am getting below WARN issue
![image](https://github.com/ing-bank/cassandra-jdbc-wrapper/assets/53363410/1450a761-8541-4273-9082-67035582730f)

**Reply by maximevw** [reply]
Hi @Arjunguptha98 
I will check that as soon as possible (probably a conflict with another module also importing this class since the "missing" class is present in the driver jar). But I think this is not blocking. Perhaps, the driver will just fail to write its internal logs but in most cases it's not really important.
I let you look for the error message on the Internet on your own in the meantime.

**Reply by maximevw** [reply]
@Arjunguptha98 
Adding this line into the `<dependencies>` tag of the module.xml file of the JDBC driver, the warning message above disappeared when I test the connection:
```xml
<module name="org.apache.logging.log4j.api"/>
```
So, the module.xml file becomes:
```xml
<module xmlns="urn:jboss:module:1.9" name="com.ing.data">
  <resources>
    <resource-root path="cassandra-jdbc-wrapper-4.11.0-bundle.jar"/>
  </resources>
  <dependencies>
    <module name="javax.api"/>
    <module name="javax.transaction.api"/>
    <module name="org.apache.logging.log4j.api"/>
  </dependencies>
</module>
```

I will update my original message https://github.com/ing-bank/cassandra-jdbc-wrapper/discussions/47#discussioncomment-7810783 to have the clean procedure.
